packages/elements/src/datetime-field/index.ts

```diff
@@ -0,0 +1,708 @@
+   * @param value ISO string date/time/datetime
+   * @returns parts
+   */
+  protected formatToParts (value: string): DateTimeFormatPart[] {
+    const date = utcParse(value);
+    return this.formatter.formatToParts(date);
+  }
+
+  /**
+   * Format value to inputValue
+   * @param value Date string
+   * @returns inputValue
+   */
+  protected toInputValue (value: string): string {
+    return value ? this.formatter.format(utcParse(value)) : '';
+  }
+
+  /**
+   * Format inputValue to value
+   * @param inputValue Date string
+   * @returns value
+   */
+  protected toValue (inputValue: string): string {
+    let value = '';
+    try {
+      value = inputValue ? this.locale.parse(inputValue, this.value || this.startDate) : '';
+    }
+    catch (error) {
+      // do nothing
+    }
+    return value;
+  }
+
+  /**
+   * On *user-interaction* set the value and notify.
+   * @param value New value
+   * @returns {void}
+   */
+  protected override setValueAndNotify (value: string): void {
+    // must be called on super to avoid re-rendering of actual value
+    if (super.value !== value) {
+      super.value = value;
+      this.notifyPropertyChange('value', value);
+    }
+  }
+
+  /**
+   * Reset error state on input
+   * @returns {void}
+   */
+  protected resetError (): void {
+    if (this.error && this.checkValidity()) {
+      this.reportValidity();
+    }
+  }
+
+  /**
+   * Returns true if an input element contains valid data.
+   * @returns true if input is valid
+   */
+  public checkValidity (): boolean {
+    const inputValue = this.inputValue;
+
+    // Invalid input value
+    if (inputValue && !this.toValue(inputValue)) {
+      return false;
+    }
+
+    const value = this.value;
+
+    // No support for required
+    if (value === '') {
+      return true;
+    }
+
+    // Value before min
+    if (this.min && value !== this.min && isBefore(value, this.min)) {
+      return false;
+    }
+
+    // Value after max
+    if (this.max && value !== this.max && isAfter(value, this.max)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Validate input. Mark as error if input is invalid
+   * @returns false if there is an error
+   */
+  public reportValidity (): boolean {
+    const hasError = !this.checkValidity();
+    this.notifyErrorChange(hasError);
+    return !hasError;
+  }
+
+  /**
+   * Select part
+   * @param index Part index
+   * @param parts The list of parts
+   * @returns {void}
+   */
+  protected selectPart (index = 0, parts: DateTimeFormatPart[]): void {
+    const { selectionStart, selectionEnd } = selectPart(index, parts);
+    this.partLabel = parts[index] ? parts[index].type : '';
+    this.setSelectionRange(selectionStart, selectionEnd);
+  }
+
+  /**
+   * Change value or selection based on keyboard navigation
+   * @param key Navigation key
+   * @param event Keyboard event
+   * @returns {void}
+   */
+  protected onNavigation (key: NavigationKeys, event: KeyboardEvent): void {
```

Comment:
Can we split this method into smaller ones, they are relatable and basically does the similar things but I feel nested switch case is a bit hard to read.